File: biomedicus-uima/src/main/java/edu/umn/biomedicus/uima/labels/UimaLabelIndex.java
/*
 * Copyright (c) 2017 Regents of the University of Minnesota.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package edu.umn.biomedicus.uima.labels;

import com.google.inject.Inject;
import edu.umn.biomedicus.framework.store.AbstractLabelIndex;
import edu.umn.biomedicus.framework.store.DefaultLabelIndex;
import edu.umn.biomedicus.framework.store.ImmutableDistinctSpanMap;
import edu.umn.biomedicus.framework.store.ImmutableSpanMap;
import edu.umn.biomedicus.framework.store.Label;
import edu.umn.biomedicus.framework.store.LabelIndex;
import edu.umn.biomedicus.framework.store.OrderedSpanMap;
import edu.umn.biomedicus.framework.store.Span;
import edu.umn.biomedicus.framework.store.SpansMap;
import edu.umn.biomedicus.framework.store.TextLocation;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import javax.annotation.Nullable;
import org.apache.uima.cas.CAS;
import org.apache.uima.cas.FSIterator;
import org.apache.uima.cas.Type;
import org.apache.uima.cas.text.AnnotationFS;
import org.apache.uima.cas.text.AnnotationIndex;

public final class UimaLabelIndex<T> extends AbstractLabelIndex<T> {

  private final CAS cas;
  private final AnnotationIndex<AnnotationFS> index;
  private final LabelAdapter<T> labelAdapter;
  private final Type annotationType;
  private final Type type;
  @Nullable
  private transient LabelIndex<T> inflated;

  @Inject
  public UimaLabelIndex(CAS cas, LabelAdapter<T> labelAdapter) {
    this.cas = cas;
    this.labelAdapter = labelAdapter;
    annotationType = cas.getTypeSystem()
        .getType("uima.tcas.Annotation");
    type = labelAdapter.getType();
    index = cas.getAnnotationIndex(type);
  }

  @Override
  public LabelIndex<T> containing(TextLocation textLocation) {
    return inflate().containing(textLocation);
  }

  @Override
  public LabelIndex<T> insideSpan(TextLocation textLocation) {
    AnnotationFS bound = cas.createAnnotation(annotationType,
        textLocation.getBegin() - 1, textLocation.getEnd() + 1);
    FSIterator<AnnotationFS> subiterator = index.subiterator(bound);
    SpansMap<T> spansMap;
    if (labelAdapter.isDistinct()) {
      ImmutableDistinctSpanMap.Builder<T> builder
          = ImmutableDistinctSpanMap.builder();
      while (subiterator.hasNext()) {
        Label<T> label = labelAdapter
            .annotationToLabel(subiterator.next());
        if (textLocation.contains(label)) {
          builder.add(label);
        }
      }
      spansMap = builder.build();
    } else {
      OrderedSpanMap<T> orderedSpanMap = new OrderedSpanMap<>();
      while (subiterator.hasNext()) {
        Label<T> label = labelAdapter
            .annotationToLabel(subiterator.next());
        if (textLocation.contains(label)) {
          orderedSpanMap.put(label, label.getValue());
        }
      }
      spansMap = new ImmutableSpanMap<>(orderedSpanMap);
    }
    return new DefaultLabelIndex<>(spansMap);
  }

  @Override
  public LabelIndex<T> leftwardsFrom(TextLocation span) {
    return inflate().leftwardsFrom(span);
  }

  @Override
  public LabelIndex<T> rightwardsFrom(TextLocation span) {
    return inflate().rightwardsFrom(span);
  }

  @Override
  public LabelIndex<T> ascendingBegin() {
    return inflate().ascendingBegin();
  }

  @Override
  public LabelIndex<T> descendingBegin() {
    return inflate().descendingBegin();
  }

  @Override
  public LabelIndex<T> ascendingEnd() {
    return inflate().ascendingEnd();
  }

  @Override
  public LabelIndex<T> descendingEnd() {
    return inflate().descendingEnd();
  }

  @Override
  public Optional<Label<T>> first() {
    FSIterator<AnnotationFS> it = index.iterator();
    if (it.hasNext()) {
      return Optional.of(labelAdapter.annotationToLabel(it.next()));
    }
    return Optional.empty();
  }

  @Override
  public Optional<Label<T>> withTextLocation(TextLocation textLocation) {
    AnnotationFS bound = cas.createAnnotation(annotationType,
        textLocation.getBegin() - 1, textLocation.getEnd() + 1);
    FSIterator<AnnotationFS> subiterator = index.subiterator(bound);
    while (subiterator.hasNext()) {
      AnnotationFS next = subiterator.next();
      if (next.getBegin() == textLocation.getBegin()
          && next.getEnd() == textLocation.getEnd()) {
        return Optional.of(labelAdapter.annotationToLabel(next));
      }
    }
    return Optional.empty();
  }

  @Override
  public Set<Span> spans() {
    return inflate().spans();
  }

  @Override
  public Collection<T> values() {
    return inflate().values();
  }

  @Override
  public List<Label<T>> asList() {
    return inflate().asList();
  }

  @Override
  public List<Span> spansAsList() {
    return inflate().spansAsList();
  }

  @Override
  public List<T> valuesAsList() {
    return inflate().valuesAsList();
  }

  @Override
  public Iterator<Label<T>> iterator() {
    return new FSIteratorAdapter<>(index, labelAdapter::annotationToLabel);
  }

  @Override
  public int size() {
    return index.size();
  }

  public LabelIndex<T> inflate() {
    if (inflated != null) {
      return inflated;
    }

    SpansMap<T> spansMap;
    if (labelAdapter.isDistinct()) {
      spansMap = ImmutableDistinctSpanMap.<T>builder()
          .addAll(this)
          .build();
    } else {
      OrderedSpanMap<T> orderedSpanMap = new OrderedSpanMap<>();
      for (Label<T> tLabel : this) {
        orderedSpanMap.put(tLabel, tLabel.getValue());
      }
      spansMap = new ImmutableSpanMap<>(orderedSpanMap);
    }
    return (inflated = new DefaultLabelIndex<>(spansMap));
  }
}

File: biomedicus-uima/src/main/kotlin/edu/umn/biomedicus/uima/labels/UimaLabelIndex.kt
/*
 * Copyright (c) 2018 Regents of the University of Minnesota.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package edu.umn.biomedicus.uima.labels

import com.google.inject.Inject
import edu.umn.nlpengine.*
import org.apache.uima.cas.CAS
import org.apache.uima.cas.text.AnnotationFS
import org.apache.uima.cas.text.AnnotationIndex


class UimaLabelIndex<T : Label> @Inject constructor(
        cas: CAS,
        private val labelAdapter: LabelAdapter<T>
) : AbstractLabelIndex<T>() {
    override val labelClass: Class<T>
        get() = labelAdapter.labelClass

    private val index: AnnotationIndex<AnnotationFS> = cas.getAnnotationIndex(labelAdapter.type)

    private val inflated: LabelIndex<T> by lazy {
        if (labelAdapter.distinct)
            DistinctLabelIndex(labelClass, this)
        else
            StandardLabelIndex(labelClass, this)
    }

    override val size: Int by lazy {
        var count = 0
        val it = iterator()
        while (it.hasNext()) {
            it.next()
            count++
        }
        count
    }

    override fun containing(startIndex: Int, endIndex: Int): LabelIndex<T> {
        return inflated.containing(startIndex, endIndex)
    }

    override fun inside(startIndex: Int, endIndex: Int): LabelIndex<T> {
        return inflated.inside(startIndex, endIndex)
    }

    override fun beginsInside(startIndex: Int, endIndex: Int): LabelIndex<T> {
        return inflated.beginsInside(startIndex, endIndex)
    }

    override fun ascendingStartIndex(): LabelIndex<T> {
        return inflated.ascendingStartIndex()
    }

    override fun descendingStartIndex(): LabelIndex<T> {
        return inflated.descendingStartIndex()
    }

    override fun ascendingEndIndex(): LabelIndex<T> {
        return inflated.ascendingEndIndex()
    }

    override fun descendingEndIndex(): LabelIndex<T> {
        return inflated.descendingEndIndex()
    }

    override fun toTheLeftOf(index: Int): LabelIndex<T> {
        return inflated.toTheLeftOf(index)
    }

    override fun toTheRightOf(index: Int): LabelIndex<T> {
        return inflated.toTheRightOf(index)
    }

    override fun first(): T? {
        val it = iterator()
        return if (!it.hasNext()) {
            null
        } else it.next()
    }

    override fun last(): T? {
        val it = index.iterator()
        it.moveToLast()
        return if (it.isValid) labelAdapter.annotationToLabel(it.get()) else null
    }

    override fun atLocation(textRange: TextRange): Collection<T> {
        return inflated.atLocation(textRange)
    }

    override fun asList(): List<T> {
        return inflated.asList()
    }

    override fun containsSpan(textRange: TextRange): Boolean {
        return inflated.containsSpan(textRange)
    }

    override fun isEmpty(): Boolean {
        return !index.iterator().hasNext()
    }

    override fun iterator(): Iterator<T> {
        val iterator = index.iterator()
        return object : Iterator<T> {
            override fun hasNext(): Boolean {
                return iterator.hasNext()
            }

            @Suppress("UNCHECKED_CAST")
            override fun next(): T {
                return labelAdapter.annotationToLabel(iterator.next())
            }
        }
    }

    override fun contains(element: T): Boolean {
        return inflated.contains(element)
    }

    override fun containsAll(elements: Collection<T>): Boolean {
        return inflated.containsAll(elements)
    }
}
